### PR TITLE
VJEPA2 model, classifier works well! need some help with fm though

### LINKED ISSUE #117: VJEPA2
## Model description
Video - Joint Embedding Predictive Architechture 2

## Open source examples / papers / references
* Paper: https://arxiv.org/abs/2506.09985
* Website: https://ai.meta.com/vjepa/
* Model code: https://github.com/huggingface/transformers/blob/main/src/transformers/models/vjepa2/modeling_vjepa2.py
* Model weights: https://huggingface.co/facebook/vjepa2-vitl-fpc64-256/tree/main

## Why this model?
* Meta's premier video foundation world model for visual understanding, planning, classification.
* Consists of a video foundation model along with a classification head (for ssv2, diving48 datasets).

### PR DESCRIPTION
**Pull request by coder0143**
Resolves #117 

**Reference**
Link: https://github.com/huggingface/transformers/blob/main/src/transformers/models/vjepa2/modeling_vjepa2.py

**Checklist**

- [x] I have read the **[Contribution Guidelines](https://github.com/jax-ml/bonsai/blob/main/CONTRIBUTING.md#contributing-a-model)** and used [pre-commit hooks](https://github.com/jax-ml/bonsai/blob/main/CONTRIBUTING.md#linting-and-type-checking) to format this commit.
- [x] I have added all the necessary **unit tests** for my change. (`run_model.py` for model usage, `test_outputs.py` and/or `model_validation_colab.ipynb` for quality).
- [x] **(If using an LLM)** I have carefully reviewed and removed all **superfluous comments** or unneeded, commented-out code. Only necessary and functional code remains.
- [x] I have signed the **[Contributor License Agreement (CLA)](https://cla.developers.google.com/about)**.

### CODE REVIEW

**Comment by coder0143** [comment]
Can anyone review my PR @jenriver @chapman20j

**Comment by chapman20j** [comment]
Hi @coder0143 . Thanks for the nice PR! This looks great. Could you please address a few comments.

File: bonsai/models/vjepa2/modeling.py [hunk 1]
```diff
@@ -0,0 +1,989 @@
+import dataclasses
+from typing import Optional, Tuple
+
+import jax
+import jax.numpy as jnp
+from flax import nnx
+from jax import Array
+
+
+@dataclasses.dataclass
```
Comment (chapman20j) [inline bonsai/models/vjepa2/modeling.py]: These dataclasses for the outputs don't add much to the implementation. Could you refactor these out to give a more minimal implementation?

File: bonsai/models/vjepa2/modeling.py [hunk 2]
```diff
@@ -0,0 +1,989 @@
...
+        self.proj = nnx.Linear(hidden_size, hidden_size, rngs=rngs)
+
+        # Grid dimensions for position encoding
+        self.grid_size = config.crop_size // config.patch_size
+        self.grid_depth = config.frames_per_clip // config.tubelet_size
+
+        # RoPE dimension splits (divide head_dim into 3 parts for d, h, w)
+        self.d_dim = int(2 * ((self.attention_head_size // 3) // 2))
+        self.h_dim = int(2 * ((self.attention_head_size // 3) // 2))
+        self.w_dim = int(2 * ((self.attention_head_size // 3) // 2))
```
Comment (chapman20j) [inline bonsai/models/vjepa2/modeling.py]: These lines are identical (up to naming). Could you condense this?

**Comment by chapman20j** [comment]
Also, you mentioned in the PR title that you "need some help with fm". Do you have a question in particular about it?

File: bonsai/models/vjepa2/modeling.py [hunk 3]
```diff
@@ -0,0 +1,989 @@
...
+    y1, y2 = y[..., 0], y[..., 1]  # Each (..., N, D/2)
+
+    # Stack as (-y2, y1) and flatten
+    y_rotated = jnp.stack((-y2, y1), axis=-1)  # (..., N, D/2, 2)
+    y_rotated = y_rotated.reshape(x.shape)  # (..., N, D)
+
+    # Apply rotation: x * cos + rotated * sin
+    rotated = (x * emb_cos) + (y_rotated * emb_sin)
+
+    return rotated
```
Comment (jenriver) [inline bonsai/models/vjepa2/modeling.py]: There are several redundancies here that span over multiple lines -- can this be rewritten into something like this?

```
def rotate_half(x: Array) -> Array:
    """Rotates half the hidden dims of the input."""
    x1, x2 = jnp.split(x, 2, axis=-1)
    return jnp.concatenate([-x2, x1], axis=-1)

def apply_rope(x: Array, pos: Array, dim: int) -> Array:
    """
    Apply 3D rotary positional embeddings.
    Simplified half-rotate implementation.
    """
    # x: (B, H, N, head_dim_fraction)
    # pos: (B, H, N)
    D = x.shape[-1]
    
    omega = 1.0 / (10000**(jnp.arange(0, D, 2, dtype=x.dtype) / D))
    angle = pos[..., None] * omega  # (B, H, N, D/2)
    
    cos = jnp.cos(angle).repeat(2, axis=-1)
    sin = jnp.sin(angle).repeat(2, axis=-1)
    
    return (x * cos) + (rotate_half(x) * sin)
```

**Comment by coder0143** [comment]
I have made the following changes:
1. Re formatted modeling.py
2. Fixed tests for pretrained models and classifiers.

The issue with foundation model was the nnx.gelu with approximation, that's why im using jax.nn.gelu.  The video processor, normalizes inputs to a narrow range that puts 54% of MLP activations in GELU's sensitive transition zone, which got carried over 24 layers causing very high atol. The classifiers work perfectly fine, and when using the fwm for downstream tasks with video processor's output, there is 0.9987 correlation with final output (~0.5 atol), to which the further tasks will be robust to.

File: bonsai/models/vjepa2/tests/test_outputs_vjepa2.py [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+        self.assertEqual(torch_hidden.shape, flax_hidden.shape)
+
+        torch.testing.assert_close(
+            torch.tensor(flax_hidden),
+            torch.tensor(torch_hidden),
+            rtol=1e-1,
+            atol=1.0,
+        )
+
+    def test_get_vision_features(self):
```
Comment (chapman20j) [inline bonsai/models/vjepa2/tests/test_outputs_vjepa2.py]: I think this tests the same thing as the previous test since `get_vision_features` runs the encoder. Could you remove this test?

File: bonsai/models/vjepa2/modeling.py [hunk 5]
```diff
@@ -0,0 +1,630 @@
...
+        self.proj = nnx.Linear(hidden_size, hidden_size, rngs=rngs)
+
+        self.grid_size = config.crop_size // config.patch_size
+        self.grid_depth = config.frames_per_clip // config.tubelet_size
+
+        # RoPE dimension splits (divide head_dim into 3 parts for d, h, w)
+        rope_dim = int(2 * ((self.attention_head_size // 3) // 2))
+        self.d_dim = rope_dim
+        self.h_dim = rope_dim
+        self.w_dim = rope_dim
```
Comment (chapman20j) [inline bonsai/models/vjepa2/modeling.py]: These are identical. Can we just update this to `self.rop_dim`?

**Comment by chapman20j** [comment]
Hi @coder0143 . I gave this another pass. The code looks nice! Could you please implement the suggested changes? Most of the changes involve simplifying some of the code to reduce the total number of lines. Also, the jit forward method is important as it lets us get lots of performance out of the code. This also improves compatibility with other libraries that leverage jit for expensive workloads. Thanks for your hard work on this!

**Comment by coder0143** [comment]
Thankyou for reviewing my PR @chapman20j , I have made the necessary changes:
1. changed params
2. updated tests
3. some parts in modeling

**Review by jenriver — APPROVED** [review]
